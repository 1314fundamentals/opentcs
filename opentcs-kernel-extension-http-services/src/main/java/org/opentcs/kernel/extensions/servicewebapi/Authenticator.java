// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
