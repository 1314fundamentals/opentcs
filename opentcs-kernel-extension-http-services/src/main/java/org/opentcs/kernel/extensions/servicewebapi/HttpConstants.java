// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
