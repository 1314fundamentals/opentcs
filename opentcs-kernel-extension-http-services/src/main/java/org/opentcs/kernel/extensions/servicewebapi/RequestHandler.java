// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
