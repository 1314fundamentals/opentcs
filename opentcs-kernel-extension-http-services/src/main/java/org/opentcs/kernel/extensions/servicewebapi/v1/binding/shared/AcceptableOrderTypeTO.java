// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.io.Serializable;
import java.util.Objects;

public class AcceptableOrderTypeTO
    implements
      Serializable {

  private final String name;
  private final int priority;

  public AcceptableOrderTypeTO(String name, int priority) {
    this.name = requireNonNull(name, "name");
    this.priority = priority;
  }

  public String getName() {
    return name;
  }

  public int getPriority() {
    return priority;
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (!(o instanceof AcceptableOrderTypeTO other)) {
      return false;
    }

    return Objects.equals(name, other.getName()) && priority == other.getPriority();
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, priority);
  }

  @Override
  public String toString() {
    return "AcceptableOrderTypeTO{"
        + "name=" + name
        + ", priority=" + priority
        + '}';
  }

}
