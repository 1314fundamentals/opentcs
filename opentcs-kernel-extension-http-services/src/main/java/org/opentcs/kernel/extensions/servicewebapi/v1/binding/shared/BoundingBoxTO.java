// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.annotation.Nonnull;

/**
 */
public class BoundingBoxTO {

  private long length;
  private long width;
  private long height;
  private CoupleTO referenceOffset;

  @JsonCreator
  public BoundingBoxTO(
      @JsonProperty(value = "length", required = true)
      long length,
      @JsonProperty(value = "width", required = true)
      long width,
      @JsonProperty(value = "height", required = true)
      long height,
      @JsonProperty(value = "referenceOffset", required = true)
      @Nonnull
      CoupleTO referenceOffset
  ) {
    this.length = length;
    this.width = width;
    this.height = height;
    this.referenceOffset = requireNonNull(referenceOffset, "referenceOffset");
  }

  public long getLength() {
    return length;
  }

  public BoundingBoxTO setLength(long length) {
    this.length = length;
    return this;
  }

  public long getWidth() {
    return width;
  }

  public BoundingBoxTO setWidth(long width) {
    this.width = width;
    return this;
  }

  public long getHeight() {
    return height;
  }

  public BoundingBoxTO setHeight(long height) {
    this.height = height;
    return this;
  }

  @Nonnull
  public CoupleTO getReferenceOffset() {
    return referenceOffset;
  }

  public BoundingBoxTO setReferenceOffset(
      @Nonnull
      CoupleTO referenceOffset
  ) {
    this.referenceOffset = requireNonNull(referenceOffset, "referenceOffset");
    return this;
  }
}
