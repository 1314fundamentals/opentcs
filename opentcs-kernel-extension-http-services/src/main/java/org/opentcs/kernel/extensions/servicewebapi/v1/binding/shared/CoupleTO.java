// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 */
public class CoupleTO {

  private long x;
  private long y;

  @JsonCreator
  public CoupleTO(
      @JsonProperty(value = "x", required = true)
      long x,
      @JsonProperty(value = "y", required = true)
      long y
  ) {
    this.x = x;
    this.y = y;
  }

  public long getX() {
    return x;
  }

  public CoupleTO setX(long x) {
    this.x = x;
    return this;
  }

  public long getY() {
    return y;
  }

  public CoupleTO setY(long y) {
    this.y = y;
    return this;
  }
}
