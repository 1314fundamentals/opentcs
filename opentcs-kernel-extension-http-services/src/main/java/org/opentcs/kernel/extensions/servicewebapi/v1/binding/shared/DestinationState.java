// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.opentcs.data.order.DriveOrder;

/**
 * A {@link org.opentcs.data.order.DriveOrder DriveOrder}'s destination.
 */
public class DestinationState {

  private String locationName = "";

  private String operation = "";

  private State state = State.PRISTINE;

  private List<Property> properties = new ArrayList<>();

  /**
   * Creates a new instance.
   */
  public DestinationState() {
  }

  @Nonnull
  public String getLocationName() {
    return locationName;
  }

  public DestinationState setLocationName(
      @Nonnull
      String name
  ) {
    this.locationName = requireNonNull(name, "name");
    return this;
  }

  @Nonnull
  public String getOperation() {
    return operation;
  }

  public DestinationState setOperation(
      @Nonnull
      String operation
  ) {
    this.operation = requireNonNull(operation, "operation");
    return this;
  }

  @Nonnull
  public State getState() {
    return state;
  }

  public DestinationState setState(
      @Nonnull
      State state
  ) {
    this.state = requireNonNull(state, "state");
    return this;
  }

  @Nonnull
  public List<Property> getProperties() {
    return properties;
  }

  public DestinationState setProperties(
      @Nonnull
      List<Property> properties
  ) {
    this.properties = requireNonNull(properties, "properties");
    return this;
  }

  public static DestinationState fromDriveOrder(DriveOrder driveOrder) {
    if (driveOrder == null) {
      return null;
    }
    DestinationState destination = new DestinationState();
    destination.setLocationName(driveOrder.getDestination().getDestination().getName());
    destination.setOperation(driveOrder.getDestination().getOperation());
    destination.setState(mapDriveOrderState(driveOrder.getState()));

    for (Map.Entry<String, String> mapEntry : driveOrder.getDestination().getProperties()
        .entrySet()) {
      destination.getProperties().add(new Property(mapEntry.getKey(), mapEntry.getValue()));
    }
    return destination;
  }

  private static DestinationState.State mapDriveOrderState(DriveOrder.State driveOrderState) {
    return switch (driveOrderState) {
      case PRISTINE -> State.PRISTINE;
      case TRAVELLING -> State.TRAVELLING;
      case OPERATING -> State.OPERATING;
      case FINISHED -> State.FINISHED;
      case FAILED -> State.FAILED;
    };
  }

  /**
   * This enumeration defines the various states a DriveOrder may be in.
   */
  public enum State {

    /**
     * A DriveOrder's initial state, indicating it being still untouched/not
     * being processed.
     */
    PRISTINE,
    /**
     * Indicates this drive order being processed at the moment.
     */
    TRAVELLING,
    /**
     * Indicates the vehicle processing an order is currently executing an
     * operation.
     */
    OPERATING,
    /**
     * Marks a DriveOrder as successfully completed.
     */
    FINISHED,
    /**
     * Marks a DriveOrder as failed.
     */
    FAILED;
  }
}
