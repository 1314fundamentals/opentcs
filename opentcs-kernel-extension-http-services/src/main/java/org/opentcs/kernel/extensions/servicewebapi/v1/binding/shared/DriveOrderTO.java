// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.util.List;

/**
 */
public class DriveOrderTO {
  private String name;
  private DestinationTO destination;
  private String transportOrder;
  private RouteTO route;
  private StateTO state;

  public DriveOrderTO(
      String name,
      DestinationTO destination,
      String transportOrder,
      RouteTO route,
      StateTO state
  ) {
    this.name = requireNonNull(name, "name");
    this.destination = requireNonNull(destination, "destination");
    this.transportOrder = requireNonNull(transportOrder, "transportOrder");
    this.route = route;
    this.state = requireNonNull(state, "state");
  }

  public String getName() {
    return name;
  }

  public DriveOrderTO setName(String name) {
    this.name = requireNonNull(name, "name");
    return this;
  }

  public DestinationTO getDestination() {
    return destination;
  }

  public DriveOrderTO setDestination(DestinationTO destination) {
    this.destination = requireNonNull(destination, "destination");
    return this;
  }

  public String getTransportOrder() {
    return transportOrder;
  }

  public DriveOrderTO setTransportOrder(String transportOrder) {
    this.transportOrder = requireNonNull(transportOrder, "transportOrder");
    return this;
  }

  public RouteTO getRoute() {
    return route;
  }

  public DriveOrderTO setRoute(RouteTO route) {
    this.route = requireNonNull(route, "route");
    return this;
  }

  public StateTO getState() {
    return state;
  }

  public DriveOrderTO setState(StateTO state) {
    this.state = requireNonNull(state, "state");
    return this;
  }

  public static class DestinationTO {

    private String destination;
    private String operation;
    private List<Property> properties;

    public DestinationTO(
        String destination,
        String operation,
        List<Property> properties
    ) {
      this.destination = requireNonNull(destination, "destination");
      this.operation = requireNonNull(operation, "operation");
      this.properties = requireNonNull(properties, "properties");
    }

    public String getDestination() {
      return destination;
    }

    public DestinationTO setDestination(String destination) {
      this.destination = requireNonNull(destination, "destination");
      return this;
    }

    public String getOperation() {
      return operation;
    }

    public DestinationTO setOperation(String operation) {
      this.operation = requireNonNull(operation, "operation");
      return this;
    }

    public List<Property> getProperties() {
      return properties;
    }

    public DestinationTO setProperties(List<Property> properties) {
      this.properties = requireNonNull(properties, "properties");
      return this;
    }
  }

  // CHECKSTYLE:OFF
  public enum StateTO {
    PRISTINE,
    TRAVELLING,
    OPERATING,
    FINISHED,
    FAILED
  }
  // CHECKSTYLE:ON
}
