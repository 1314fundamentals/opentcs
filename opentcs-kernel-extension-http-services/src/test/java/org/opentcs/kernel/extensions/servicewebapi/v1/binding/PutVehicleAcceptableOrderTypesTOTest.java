// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.AcceptableOrderTypeTO;

/**
 * Tests for {@link PutVehicleAcceptableOrderTypesTO}.
 */
class PutVehicleAcceptableOrderTypesTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  void jsonSample() {
    PutVehicleAcceptableOrderTypesTO to
        = new PutVehicleAcceptableOrderTypesTO(
            List.of(
                new AcceptableOrderTypeTO("some-orderType", 0),
                new AcceptableOrderTypeTO("another-orderType", 1),
                new AcceptableOrderTypeTO("orderType-3", 2)
            )
        );

    Approvals.verify(jsonBinder.toJson(to));
  }
}
