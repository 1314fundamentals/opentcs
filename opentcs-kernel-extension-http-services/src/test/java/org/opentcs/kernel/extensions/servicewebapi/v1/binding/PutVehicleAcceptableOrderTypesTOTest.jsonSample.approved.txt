{
  "acceptableOrderTypes" : [ {
    "name" : "some-orderType",
    "priority" : 0
  }, {
    "name" : "another-orderType",
    "priority" : 1
  }, {
    "name" : "orderType-3",
    "priority" : 2
  } ]
}
